### PR TITLE
update metrics collect to nebula 3.0

### PR DESCRIPTION
**Pull request by kqzh**
- [x] support collect space metric
- [x] support collect nebula 3.0 metric
- [x] remove bare-metal namespace label

### CODE REVIEW

File: exporter/convert.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	"strings"
+)
+
+type MetricValue struct {
+	Value string
+	Label map[string]string
+}
+
+func convertToMap(metrics []string) map[string]MetricValue {
+	matches := make(map[string]MetricValue)
```
Comment (veezhang) [inline exporter/convert.go:13]: give a default length of `len(metrics)`?

File: exporter/convert.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package exporter
+
+import (
+	"strings"
+)
+
+type MetricValue struct {
+	Value string
+	Label map[string]string
```
Comment (veezhang) [inline exporter/convert.go:9]: Rename to `Labels`?

File: exporter/convert.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	return true, seps[1]
+}
+
+// split slow_query_latency_us{space=nba}.p95.5=0 => slow_query_latency_us.p95.5=0, map[space:nba]
+func splitMetric(metric string) (string, map[string]string) {
+	start := strings.Index(metric, "{")
+	if start == -1 {
+		return metric, nil
+	}
+	end := strings.LastIndex(metric, "}")
```
Comment (veezhang) [inline exporter/convert.go:71]: if `end` is `-1`, will panic?

File: exporter/type.go [hunk 4]
```diff
@@ -1,14 +1,26 @@
...
-const DefaultClusterName = "default"
+const (
+	DefaultClusterName = "default"
+
+	ComponentLabelKey = "app.kubernetes.io/component"
+	ClusterLabelKey   = "app.kubernetes.io/cluster"
+
+	GraphdComponent   ComponentType = "graphd"
+	MetadComponent    ComponentType = "metad"
+	StoragedComponent ComponentType = "storaged"
```
Comment (veezhang) [inline exporter/type.go:11]: remove it if no more used.

File: exporter/metric.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	"strings"
+	"time"
+)
+
+func getNebulaMetrics(ipAddress string, port int32) ([]string, error) {
+	httpClient := http.Client{
+		Timeout: time.Second * 2,
+	}
+
+	resp, err := httpClient.Get(fmt.Sprintf("http://%s:%d/stats?stats=", ipAddress, port))
```
Comment (veezhang) [inline exporter/metric.go:17]: remove `?stats=`?

File: exporter/metric.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	var status nebulaStatus
+	if err := json.Unmarshal(bytes, &status); err != nil {
+		return nil, err
+	}
+
+	countStr := "count=0"
+	if status.Status == "running" {
+		countStr = strings.Replace(countStr, "0", "1", 1)
+	}
+	statusMetrics := []string{countStr}
```
Comment (veezhang) [inline exporter/metric.go:63]: It's no good to use`strings.Replace(countStr, "0", "1", 1)`?

File: exporter/exporter.go [hunk 7]
```diff
@@ -299,31 +66,25 @@ func NewNebulaExporter(ns, cluster, listenAddr string, client *kubernetes.Client
...
-	for _, metrics := range exporter.metricMap {
-		ch <- metrics.desc
-	}
-
-	klog.Info("Describe Nebula Metrics Done")
+func (exporter *NebulaExporter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	exporter.mux.ServeHTTP(w, r)
 }
 
+func (exporter *NebulaExporter) Describe(ch chan<- *prometheus.Desc) {}
```
Comment (veezhang) [inline exporter/exporter.go:86]: No more used ?

File: exporter/exporter.go [hunk 8]
```diff
@@ -349,79 +110,67 @@ func (exporter *NebulaExporter) CollectMetrics(
...
-			}
-		}
-		return
-	}
-
-	// query metrics
 	matches := convertToMap(metrics)
 	for metric, value := range matches {
-		v, err := strconv.ParseFloat(value, 64)
+		v, err := strconv.ParseFloat(value.Value, 64)
```
Comment (veezhang) [inline exporter/exporter.go:115]: It's all should be int?

File: exporter/exporter.go [hunk 9]
```diff
@@ -349,79 +110,67 @@ func (exporter *NebulaExporter) CollectMetrics(
...
 		}
+
+		ch <- prometheus.MustNewConstMetric(prometheus.NewDesc(
+			fmt.Sprintf("%s_%s_%s", FQNamespace, componentType, strings.ReplaceAll(metric, ".", "_")),
+			"", labels,
+			nil),
+			prometheus.GaugeValue,
+			v,
+			labelValues...,
+		)
```
Comment (veezhang) [inline exporter/exporter.go:140]: format it?

**Review by veezhang — APPROVED** [review]
LGTM

File: exporter/convert.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+func splitMetric(metric string) (string, map[string]string) {
+	start := strings.Index(metric, "{")
+	end := strings.LastIndex(metric, "}")
+
+	if start == -1 || end == -1 {
+		return metric, nil
+	}
+
+	label := make(map[string]string)
+	labelsStr := strings.Split(metric[start+1:end], ",")
```
Comment (MegaByte875) [inline exporter/convert.go:42]: why the sep is ","

File: exporter/exporter.go [hunk 11]
```diff
@@ -349,79 +110,70 @@ func (exporter *NebulaExporter) CollectMetrics(
...
 	}
 }
 
 func (exporter *NebulaExporter) collect(wg *sync.WaitGroup, namespace, clusterName string, instance Instance, ch chan<- prometheus.Metric) {
 	podIpAddress := instance.EndpointIP
 	podHttpPort := instance.EndpointPort
 
-	if podHttpPort == 0 {
+	if podHttpPort <= 0 {
```
Comment (MegaByte875) [inline exporter/exporter.go:151]: <0 means?

File: exporter/type.go [hunk 12]
```diff
@@ -1,14 +1,20 @@
...
-const DefaultClusterName = "default"
+const (
+	DefaultClusterName = "default"
+
+	ComponentLabelKey = "app.kubernetes.io/component"
+	ClusterLabelKey   = "app.kubernetes.io/cluster"
+
+	// FQNamespace represents the prometheus FQName
+	FQNamespace  = "nebula"
+	NonNamespace = "none_namespace"
```
Comment (MegaByte875) [inline exporter/type.go:11]: none seems more concise

File: exporter/convert.go (hunk 10 above)
Comment (kqzh) [reply exporter/convert.go:42]: becase the metric may like this `num_queries{space=nba,partition=1}`

File: exporter/exporter.go (hunk 11 above)
Comment (kqzh) [reply exporter/exporter.go:151]: because port < 0 also invalid

File: exporter/type.go (hunk 12 above)
Comment (kqzh) [reply exporter/type.go:11]: I consider concise name may  easy to used by user, so I use a long name `none_namespace`

**Review by MegaByte875 — APPROVED** [review]
LGTM

**Pull request merged**